File: services/identity-service/src/main/java/br/com/deltaglobalbank/identity/features/apiClients/listClients/ListClientsUseCase.java
package br.com.deltaglobalbank.identity.features.apiClients.listClients;

import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;

import br.com.deltaglobalbank.identity.domain.role.Role;
import br.com.deltaglobalbank.identity.domain.role.RoleRepository;
import br.com.deltaglobalbank.identity.infrastructure.persistence.entities.ApiClientEntity;
import br.com.deltaglobalbank.identity.infrastructure.persistence.entities.ApiClientRoleEntity;
import br.com.deltaglobalbank.identity.infrastructure.persistence.entities.TenantEntity;
import br.com.deltaglobalbank.identity.infrastructure.persistence.repositories.ApiKeyActiveCount;
import br.com.deltaglobalbank.identity.infrastructure.persistence.repositories.JpaApiClientRepository;
import br.com.deltaglobalbank.identity.infrastructure.persistence.repositories.JpaApiClientRoleRepository;
import br.com.deltaglobalbank.identity.infrastructure.persistence.repositories.JpaApiKeyRepository;
import br.com.deltaglobalbank.identity.infrastructure.persistence.repositories.JpaTenantRepository;
import jakarta.transaction.Transactional;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

@Service
public class ListClientsUseCase {

    public static final int MAX_PAGE_SIZE = 100;

    private final JpaApiClientRepository jpaApiClientRepository;
    private final JpaTenantRepository tenantRepository;
    private final JpaApiClientRoleRepository jpaApiClientRoleRepository;
    private final RoleRepository roleRepository;
    private final JpaApiKeyRepository jpaApiKeyRepository;

    public ListClientsUseCase(
        JpaApiClientRepository jpaApiClientRepository,
        JpaTenantRepository tenantRepository,
        JpaApiClientRoleRepository jpaApiClientRoleRepository,
        RoleRepository roleRepository,
        JpaApiKeyRepository jpaApiKeyRepository
    ) {
        this.jpaApiClientRepository = jpaApiClientRepository;
        this.tenantRepository = tenantRepository;
        this.jpaApiClientRoleRepository = jpaApiClientRoleRepository;
        this.roleRepository = roleRepository;
        this.jpaApiKeyRepository = jpaApiKeyRepository;
    }

    @Transactional
    public ListClientsResponse listClients(UUID tenantId, int page, int size) {
        int safePage = Math.max(page, 0);
        int safeSize = Math.min(Math.max(size, 1), MAX_PAGE_SIZE);
        PageRequest pageable = PageRequest.of(safePage, safeSize, Sort.Direction.DESC, "createdAt");

        Page<ApiClientEntity> clientsPage = jpaApiClientRepository.findByTenantId(tenantId, pageable);
        if (clientsPage.getContent().isEmpty()) {
            return new ListClientsResponse(List.of(), safePage, safeSize, 0, 0);
        }

        Set<UUID> clientIds = clientsPage.getContent().stream().map(ApiClientEntity::getId).collect(Collectors.toSet());
        TenantEntity tenant = tenantRepository.findById(tenantId).orElse(null);

        Map<UUID, List<ApiClientRoleEntity>> rolesByClientId = jpaApiClientRoleRepository.findAllByApiClientIdIn(clientIds)
            .stream()
            .collect(Collectors.groupingBy(ApiClientRoleEntity::getApiClientId));

        Set<UUID> roleIds = rolesByClientId.values().stream()
            .flatMap(List::stream)
            .map(ApiClientRoleEntity::getRoleId)
            .collect(Collectors.toSet());
        Map<UUID, Role> rolesById = roleIds.isEmpty()
            ? Map.of()
            : roleRepository.findAllByIds(roleIds).stream().collect(Collectors.toMap(Role::getId, r -> r));

        Instant now = Instant.now();

        Map<UUID, Long> activeCountByClient = jpaApiKeyRepository.countActiveByApiClientIdIn(clientIds, now).stream()
            .collect(Collectors.toMap(ApiKeyActiveCount::getApiClientId, ApiKeyActiveCount::getTotal));

        List<ItemsResponse> items = clientsPage.getContent().stream()
            .map(client -> {
                List<String> roleCodes = rolesByClientId.getOrDefault(client.getId(), List.of()).stream()
                    .map(it -> rolesById.get(it.getRoleId()))
                    .filter(java.util.Objects::nonNull)
                    .map(r -> r.getCode().toString())
                    .toList();
                int activeKeys = activeCountByClient.getOrDefault(client.getId(), 0L).intValue();

                return new ItemsResponse(
                    client.getId(),
                    client.getTenantId(),
                    tenant != null ? tenant.getSlug() : "unknown",
                    client.getName(),
                    client.getDescription() != null ? client.getDescription() : "",
                    client.getStatus(),
                    roleCodes,
                    activeKeys,
                    client.getCreatedAt()
                );
            })
            .toList();

        return new ListClientsResponse(
            items,
            safePage,
            safeSize,
            clientsPage.getTotalElements(),
            clientsPage.getTotalPages()
        );
    }
}

File: services/identity-service/src/main/kotlin/br/com/deltaglobalbank/identity/features/apiClients/listClients/ListClientsUseCase.kt
package br.com.deltaglobalbank.identity.features.apiClients.listClients

import br.com.deltaglobalbank.identity.domain.role.Role
import br.com.deltaglobalbank.identity.domain.role.RoleRepository
import br.com.deltaglobalbank.identity.infrastructure.persistence.repositories.JpaApiClientRepository
import br.com.deltaglobalbank.identity.infrastructure.persistence.repositories.JpaApiClientRoleRepository
import br.com.deltaglobalbank.identity.infrastructure.persistence.repositories.JpaApiKeyRepository
import br.com.deltaglobalbank.identity.infrastructure.persistence.repositories.JpaTenantRepository
import jakarta.transaction.Transactional
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import java.time.Instant
import java.util.UUID

@Service
class ListClientsUseCase(private val jpaApiClientRepository: JpaApiClientRepository, private val tenantRepository: JpaTenantRepository, private val jpaApiClientRoleRepository : JpaApiClientRoleRepository, private val roleRepository: RoleRepository, private val jpaApiKeyRepository : JpaApiKeyRepository) {

    companion object { const val MAX_PAGE_SIZE = 100}

    @Transactional
    fun listClients(tenantId: UUID, page: Int, size: Int): ListClientsResponse {

        val safePage = page.coerceAtLeast(0)
        val safeSize = size.coerceIn(1, MAX_PAGE_SIZE)
        val pageable = PageRequest.of(safePage, safeSize, Sort.Direction.DESC, "createdAt")

        val clientsPage = jpaApiClientRepository.findByTenantId(tenantId, pageable)
        if (clientsPage.content.isEmpty()) return ListClientsResponse(emptyList(), safePage, safeSize, 0, 0)

        val clientIds = clientsPage.content.map { it.id }.toSet()
        val tenant = tenantRepository.findById(tenantId).orElse(null)

        val rolesByClientId = jpaApiClientRoleRepository.findAllByApiClientIdIn(clientIds)
            .groupBy { it.apiClientId }

        val roleIds = rolesByClientId.values.flatten().map { it.roleId }.toSet()
        val rolesById: Map<UUID, Role> =
            if (roleIds.isEmpty()) emptyMap() else roleRepository.findAllByIds(roleIds).associateBy { it.id }

        val now = Instant.now()

        val activeCountByClient = jpaApiKeyRepository
            .countActiveByApiClientIdIn(clientIds, now)
            .associate { it.apiClientId to it.total }

        val items = clientsPage.content.map { client ->
            val roleCodes = rolesByClientId[client.id].orEmpty()
                .mapNotNull { rolesById[it.roleId]?.code?.toString() }
            val activeKeys = (activeCountByClient[client.id] ?: 0L).toInt()

            ItemsResponse(
                id = client.id,
                tenantId = client.tenantId,
                tenantSlug = tenant?.slug ?: "unknown",
                name = client.name,
                description = client.description ?: "",
                status = client.status,
                roles = roleCodes,
                activeKeysCount = activeKeys,
                createdAt = client.createdAt
            )
        }

        return ListClientsResponse(
            items = items,
            page = safePage,
            size = safeSize,
            totalElements = clientsPage.totalElements,
            totalPages = clientsPage.totalPages,
        )
    }


}